### PR TITLE
feat(p2p/grpc): creating gRPC behaviour

### PR DESCRIPTION
**Pull request by Freyskeyd**
https://img.shields.io/badge/1%20PR-grpc%20over%20p2p-blue 

# Description 
Next PR: #332 

This PR adds the gRPC behaviour in order to use gRPC client/server with p2p TCP connections.

The goal is to be able to rely on protobuf schema to dictate our network behaviours, allowing any other services that use gRPC to communicate with us transparently.
It also prevents us to generate and deal with complex custom-made Client/Server.
It also eases the `request/response` between two nodes and encapsulate the logic into dedicated structs

## PR Checklist:

- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added or updated tests that comprehensively prove my change is effective or that my feature works

### CODE REVIEW

File: crates/topos-p2p/src/behaviour/grpc.rs [hunk 1]
```diff
@@ -0,0 +1,516 @@
...
+        };
+
+        let connection = Connection {
+            id: connection_id,
+            address,
+            request_id: None,
+            channel: None,
+        };
+
+        self.connected.entry(peer_id).or_default().push(connection);
```
Comment (dvdplm) [inline crates/topos-p2p/src/behaviour/grpc.rs:244]: Would it work to use `or_insert_with()` here too?

File: crates/topos-p2p/src/behaviour/grpc.rs [hunk 2]
```diff
@@ -0,0 +1,516 @@
...
+    /// Handle the [`ConnectionEstablished`] event comming from the [`Swarm`]
+    /// and try to open a gRPC channel using a [`ConnectionHandler`].
+    fn on_connection_established(
+        &mut self,
+        ConnectionEstablished {
+            peer_id,
+            connection_id,
+            endpoint,
+            failed_addresses,
+            other_established,
```
Comment (dvdplm) [inline crates/topos-p2p/src/behaviour/grpc.rs:229]: I'm not sure I understand what this value means. Other connections with this same peer?

File: crates/topos-p2p/src/behaviour/grpc.rs [hunk 3]
```diff
@@ -0,0 +1,516 @@
...
+    fn try_connect(&mut self, peer_id: &PeerId) {
+        if let Some(connections) = self.connected.get_mut(peer_id) {
+            let connection = connections.first_mut();
+            if let Some(connection) = connection {
+                if let Some(OutboundConnectionRequest {
+                    request_id,
+                    notifier,
+                }) = self.pending_outbound_connections.get(peer_id)
+                {
+                    info!("gRPC Outbound connection established with {peer_id}");
```
Comment (dvdplm) [inline crates/topos-p2p/src/behaviour/grpc.rs:334]: Maybe downgrade this to a `debug!`?

File: crates/topos-p2p/src/behaviour/grpc.rs [hunk 4]
```diff
@@ -0,0 +1,516 @@
...
+            request_id: None,
+            channel: None,
+        };
+
+        self.connected.entry(peer_id).or_default().push(connection);
+
+        // TODO refactor this to better handle connection
+        // If there is no current established connection it means that it's the
+        // first connection with that peer
+        if other_established == 0 {
```
Comment (dvdplm) [inline crates/topos-p2p/src/behaviour/grpc.rs:249]: I guess that what I don't understand is under which conditions we could get a `ConnectionEstablished` after we have established the first connection?

File: crates/topos-p2p/src/behaviour/grpc.rs [hunk 5]
```diff
@@ -0,0 +1,516 @@
...
+            connection_id,
+            endpoint,
+            failed_addresses,
+            other_established,
+        }: ConnectionEstablished,
+    ) {
+        let address = match endpoint {
+            ConnectedPoint::Dialer { address, .. } => Some(address.clone()),
+            ConnectedPoint::Listener { .. } => None,
+        };
```
Comment (dvdplm) [inline crates/topos-p2p/src/behaviour/grpc.rs:235]: SO if we have an address then we called them, otherwise they called us, is that correct? Do we store the address in our `Connection` so we can re-dial? Or do we need it for other reasons?

File: crates/topos-p2p/src/behaviour/grpc.rs [hunk 6]
```diff
@@ -0,0 +1,516 @@
...
+            self.try_connect(&peer_id);
+        }
+    }
+
+    /// Starts the gRPC service if not already started
+    fn on_new_listener(&mut self, listener_id: ListenerId) {
+        if let Some(service) = self.service.take() {
+            let (tx, rx) = mpsc::unbounded_channel();
+            self.inbound_stream = Some(tx);
+            // TODO: Switch to serve_with_incoming_shutdown at some point
```
Comment (dvdplm) [inline crates/topos-p2p/src/behaviour/grpc.rs:259]: Would be good to have a ticket for this.

File: crates/topos-p2p/src/behaviour/grpc.rs [hunk 7]
```diff
@@ -0,0 +1,516 @@
...
+            tokio::spawn(service.serve_with_incoming(proxy::GrpcProxy::new(rx)));
+            info!("New gRPC proxy start and listening on {listener_id:?}");
```
Comment (dvdplm) [inline crates/topos-p2p/src/behaviour/grpc.rs:261]:
```suggestion
            info!("New gRPC proxy started and listening on {listener_id:?}");
```

File: crates/topos-p2p/src/behaviour/grpc.rs [hunk 8]
```diff
@@ -0,0 +1,516 @@
...
+        } else {
+            warn!("Tried to instantiate gRPC proxy but service is missing");
```
Comment (dvdplm) [inline crates/topos-p2p/src/behaviour/grpc.rs:263]:
```suggestion
            warn!("Tried to instantiate a gRPC proxy on {listener_id:?} but the service is missing");
```

File: crates/topos-p2p/src/behaviour/grpc.rs [hunk 9]
```diff
@@ -0,0 +1,516 @@
...
+        &mut self,
+        ConnectionClosed {
+            peer_id,
+            connection_id,
+            endpoint,
+            handler,
+            remaining_established,
+        }: ConnectionClosed<<Self as NetworkBehaviour>::ConnectionHandler>,
+    ) {
+        warn!("Connection {connection_id} closed with peer {peer_id}");
```
Comment (dvdplm) [inline crates/topos-p2p/src/behaviour/grpc.rs:278]: This is normal behaviour right? If yes, then `debug!` is more appropriate.

File: crates/topos-p2p/src/behaviour/grpc.rs [hunk 10]
```diff
@@ -0,0 +1,516 @@
...
+    /// Handle the [`DialFailure`] event comming from the [`Swarm`]
+    fn on_dial_failure(
+        &mut self,
+        DialFailure {
+            peer_id,
+            error,
+            connection_id,
+        }: DialFailure,
+    ) {
+        println!("Dial failure {:?}", error);
```
Comment (dvdplm) [inline crates/topos-p2p/src/behaviour/grpc.rs:296]:
```suggestion
        warn!("Dial failure {:?}", error);
```

File: crates/topos-p2p/src/behaviour/grpc.rs [hunk 11]
```diff
@@ -0,0 +1,516 @@
...
+                    {
+                        self.pending_events.push_back(ToSwarm::GenerateEvent(
+                            Event::OutboundFailure {
+                                peer_id,
+                                request_id,
+                                error: OutboundError::DialFailure,
+                            },
+                        ));
+
+                        let _ = notifier.send(Err(OutboundError::DialFailure));
```
Comment (dvdplm) [inline crates/topos-p2p/src/behaviour/grpc.rs:316]: It looks like we're loose track of what the error causing the dial failure. I'm not what happens next here, but whatever code is handling `Event::OutboundFailure` and `OutboundError::DialFailure` will not have access to the error. Is that what we want?

File: crates/topos-p2p/src/behaviour/grpc.rs [hunk 12]
```diff
@@ -0,0 +1,516 @@
...
+        // Sending event to both `Swarm` and `ConnectionHandler`
+        if let Some(ev) = self.pending_events.pop_front() {
+            return Poll::Ready(ev);
+        }
+
+        // When channel has been negotiated by the [`ConnectionHandler`] we need
+        // to update the [`Connection`] with the channel.
+        match self.pending_negotiated_channels.poll_next_unpin(cx) {
+            Poll::Ready(Some((Ok(channel), request_id, peer_id))) => {
+                tracing::debug!("gRPC channel ready for {} {}", peer_id, request_id);
```
Comment (dvdplm) [inline crates/topos-p2p/src/behaviour/grpc.rs:466]:
```suggestion
                debug!("gRPC channel ready for {} {}", peer_id, request_id);
```

File: crates/topos-p2p/src/behaviour/grpc/stream.rs [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+
+impl Future for NegotiatedGrpcOutboundStream {
+    type Output = Result<GrpcStream, BoxError>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let fut = self.project();
+        let x = fut
+            .stream_rx
+            .lock()
+            .unwrap()
```
Comment (dvdplm) [inline crates/topos-p2p/src/behaviour/grpc/stream.rs:51]: I know of the stdlib's mutexe poisoning stuff, but I think we should use `expect("why it's reasonable to make the assumptions we make")` as a pattern in all our code. It amounts to the same thing, just a wee bit clearer to the reader that we know what we're doing?

File: crates/topos-p2p/src/behaviour/grpc/event.rs [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+    OutboundSuccess {
+        peer_id: PeerId,
+        request_id: RequestId,
+        channel: Channel,
+    },
+
+    // InboundNegotiatedStream {
+    //     request_id: RequestId,
+    //     stream: libp2p::Stream,
+    // },
```
Comment (dvdplm) [inline crates/topos-p2p/src/behaviour/grpc/event.rs:23]: Needed for later PRs or left-overs?

File: crates/topos-p2p/src/behaviour/grpc/event.rs [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+    InboundNegotiatedConnection {
+        request_id: RequestId,
+        connection_id: ConnectionId,
+    },
+
+    // OutboundNegotiatedStream {
+    //     request_id: RequestId,
+    //     stream: libp2p::Stream,
+    // },
```
Comment (dvdplm) [inline crates/topos-p2p/src/behaviour/grpc/event.rs:32]: Needed for later PRs or left-overs?

File: crates/topos-p2p/src/behaviour/grpc/connection.rs [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+
+/// Connection struct which represent a connection between two nodes
+/// It contains the connection id, the address of the node, the request id
+/// and the gRPC channel which is used to communicate with the node.
+#[derive(Debug)]
+pub(crate) struct Connection {
+    /// The connection id
+    pub(crate) id: ConnectionId,
+
+    /// The address of the node
```
Comment (dvdplm) [inline crates/topos-p2p/src/behaviour/grpc/connection.rs:21]: THe address of the **remote** node right?

File: crates/topos-p2p/src/behaviour/grpc/connection.rs [hunk 17]
```diff
@@ -0,0 +1,90 @@
...
+#[derive(Debug)]
+pub struct OutboundConnectedConnection {
+    #[allow(dead_code)]
+    pub(crate) request_id: RequestId,
+    // TODO: Remove unused when gRPC behaviour is activated
+    #[allow(unused)]
+    pub(crate) channel: tonic::transport::Channel,
+}
+
+/// Enum which represent the different state of an outbound connection
```
Comment (dvdplm) [inline crates/topos-p2p/src/behaviour/grpc/connection.rs:47]:
```suggestion
/// Enum which represent the different states of an outbound connection
```

**Review by dvdplm — APPROVED** [review]
This is a big PR, good job! Can't say I understood the whole thing fully, but overall I think it's looking good. I'm approving this, but it would be good if you addressed the various nitpicks. The one true concern I have is the possibility of race conditions on `RequestId`.

File: crates/topos-p2p/Cargo.toml [hunk 18]
```diff
@@ -15,15 +15,30 @@ prometheus-client.workspace = true
...
 hex.workspace = true
 unsigned-varint = { version = "0.7.1", features = ["futures", "asynchronous_codec"] }
 void = "1"
+tonic = {workspace = true, features = ["tls", "tls-roots"]}
+tower.workspace = true
 
+topos-api = {path = "../topos-api/", features = ["grpc"]}
 topos-metrics = { path = "../topos-metrics/" }
 topos-crypto = { path = "../topos-crypto" }
+smallvec = "1.11.1"
```
[the remark this answers is not in the document]
Comment (Freyskeyd) [reply crates/topos-p2p/Cargo.toml:31]: Tinyvec expects that the element type implement `Default` which is not the case for `Connection`

File: crates/topos-p2p/src/behaviour/grpc.rs [hunk 19]
```diff
@@ -0,0 +1,516 @@
...
+/// to a peer without having a gRPC service to expose.
+pub(crate) struct Behaviour {
+    /// The optional gRPC service to expose
+    service: Option<Router>,
+    /// The next request id to use
+    next_request_id: RequestId,
+    /// The next inbound request id to use
+    next_inbound_request_id: Arc<AtomicU64>,
+    /// The list of connected peers with the associated gRPC channel
+    connected: HashMap<PeerId, SmallVec<[Connection; 2]>>,
```
[the remark this answers is not in the document]
Comment (Freyskeyd) [reply crates/topos-p2p/src/behaviour/grpc.rs:84]: 2 connections because we could have one `inbound` and one `outound`.
6 `multiaddr` because of the different transports (dns/tcp/ws/relay...)

File: crates/topos-p2p/src/behaviour/grpc.rs (hunk 2 above)
Comment (Freyskeyd) [reply crates/topos-p2p/src/behaviour/grpc.rs:229]: Yes, it's the number of existing established connection with this peer

File: crates/topos-p2p/src/behaviour/grpc.rs (hunk 5 above)
Comment (Freyskeyd) [reply crates/topos-p2p/src/behaviour/grpc.rs:235]: The behaviour need to expose addresses in order to connect or reconnect

File: crates/topos-p2p/src/behaviour/grpc.rs (hunk 1 above)
Comment (Freyskeyd) [reply crates/topos-p2p/src/behaviour/grpc.rs:244]: `or_insert_with` would need to provide an empty `vec` and then `push` to it?
It is more clear and recommanded to rely on `Default`

File: crates/topos-p2p/src/behaviour/grpc.rs (hunk 4 above)
Comment (Freyskeyd) [reply crates/topos-p2p/src/behaviour/grpc.rs:249]: If a node exposes 10 addresses, you can have 10 different connections for the same peer. Or even multiple connections per transport in order to avoid HOL blocking.

File: crates/topos-p2p/src/behaviour/grpc.rs [hunk 20]
```diff
@@ -0,0 +1,516 @@
...
+        } else {
+            debug!("Buffering sender as no available connection to peer {peer_id} yet");
+            self.open_connection(peer_id)
+        }
+    }
+
+    /// Return the next outbound request id
+    fn next_request_id(&mut self) -> RequestId {
+        let request_id = self.next_request_id;
+        self.next_request_id.0 += 1;
```
[the remark this answers is not in the document]
Comment (Freyskeyd) [reply crates/topos-p2p/src/behaviour/grpc.rs:191]: > Isn't this racy?

I don't think I understand correctly, can you rephrase?

File: crates/topos-p2p/src/behaviour/grpc.rs (hunk 11 above)
Comment (Freyskeyd) [reply crates/topos-p2p/src/behaviour/grpc.rs:316]: We don't really care about the reason of this `DialFailure`, we only care about the `condition` failure.

File: crates/topos-p2p/src/behaviour/grpc/connection.rs (hunk 16 above)
Comment (Freyskeyd) [reply crates/topos-p2p/src/behaviour/grpc/connection.rs:21]: It's the address of the node for this connection

File: crates/topos-p2p/src/behaviour/grpc.rs (hunk 20 above)
[the remark this answers is not in the document]
Comment (Freyskeyd) [reply crates/topos-p2p/src/behaviour/grpc.rs:191]: If it's about race condition, because we have mutable reference on `self` we can't have race condition.

File: crates/topos-p2p/src/behaviour/grpc.rs (hunk 4 above)
Comment (dvdplm) [reply crates/topos-p2p/src/behaviour/grpc.rs:249]: 💡 aha!

File: crates/topos-p2p/src/behaviour/grpc.rs (hunk 11 above)
Comment (dvdplm) [reply crates/topos-p2p/src/behaviour/grpc.rs:316]: 🤷‍♂️ Fair enough.

File: crates/topos-p2p/src/behaviour/grpc/connection.rs (hunk 16 above)
Comment (dvdplm) [reply crates/topos-p2p/src/behaviour/grpc/connection.rs:21]: Hmm, so it's the address of whichever node dialed the other?

File: crates/topos-p2p/src/behaviour/grpc/connection.rs (hunk 17 above)
Comment (dvdplm) [reply crates/topos-p2p/src/behaviour/grpc/connection.rs:47]:
```suggestion
/// Enum that represents the different states of an outbound connection
```

File: crates/topos-p2p/src/behaviour/grpc.rs (hunk 19 above)
[the remark this answers is not in the document]
Comment (dvdplm) [reply crates/topos-p2p/src/behaviour/grpc.rs:84]: Maybe update the doc comment with this info?

But why is it a growable vector if we know it's always a fixed size?

File: crates/topos-p2p/src/behaviour/grpc.rs (hunk 1 above)
Comment (dvdplm) [reply crates/topos-p2p/src/behaviour/grpc.rs:244]: No, you'd build the content right away so no push needed. `self.connected.entry(peer_id).or_insert_with(|| vec![connection])` works I think.

I don't think it matters, but as a reader, the `or_default()` forced me to go look up what the default is, which isn't relevant in this case.

Comment (Freyskeyd) [reply crates/topos-p2p/src/behaviour/grpc.rs:244]: yeah but the behaviour is different and doesn't cover what is expected. Look this playground: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=10a11a2cea226ce8a965d1e0a58ad9ae

File: crates/topos-p2p/src/behaviour/grpc.rs (hunk 19 above)
[the remark this answers is not in the document]
Comment (Freyskeyd) [reply crates/topos-p2p/src/behaviour/grpc.rs:84]: We don't know it's always a fixed size. We know that it is common to have 2 connection but we may have more, that the point of using `SmallVec`.

File: crates/topos-p2p/src/behaviour/grpc/connection.rs (hunk 16 above)
Comment (Freyskeyd) [reply crates/topos-p2p/src/behaviour/grpc/connection.rs:21]: Yes the address of the node represented by the connection

File: crates/topos-p2p/src/tests/behaviour/grpc.rs [hunk 21]
```diff
@@ -111,7 +111,7 @@ async fn opening_outbound_stream() {}
 async fn opening_outbound_stream_half_close() {}
 
 #[test(tokio::test)]
-#[ignore = "Need to find a way to properly close the connection after sending the query"]
+#[ignore = "TP-757: Need to find a way to properly close the connection after sending the query"]
```
Comment (dvdplm) [inline crates/topos-p2p/src/tests/behaviour/grpc.rs:114]: ❤️
